storage/jellyfish_merkle/src/node_type/node_type_test.rs

```diff
@@ -669,3 +669,162 @@ fn test_internal_hash_and_proof() {
         }
     }
 }
+
+enum BinaryTreeNode {
+    Internal(BinaryTreeInternalNode),
+    Child(BinaryTreeChildNode),
+    Null,
+}
+
+impl BinaryTreeNode {
+    fn new_child(index: u8, child: &Child) -> Self {
+        Self::Child(BinaryTreeChildNode {
+            index,
+            version: child.version,
+            hash: child.hash,
+            is_leaf: child.is_leaf,
+        })
+    }
+
+    fn new_internal(
+        first_child_index: u8,
+        num_children: u8,
+        left: BinaryTreeNode,
+        right: BinaryTreeNode,
+    ) -> Self {
+        let hash = SparseMerkleInternalNode::new(left.hash(), right.hash()).hash();
+
+        Self::Internal(BinaryTreeInternalNode {
+            first_child_index,
+            num_children,
+            left: Rc::new(left),
+            right: Rc::new(right),
+            hash,
+        })
+    }
+
+    fn hash(&self) -> HashValue {
+        match self {
+            BinaryTreeNode::Internal(node) => node.hash,
+            BinaryTreeNode::Child(node) => node.hash,
+            BinaryTreeNode::Null => *SPARSE_MERKLE_PLACEHOLDER_HASH,
+        }
+    }
+}
+
+struct BinaryTreeInternalNode {
+    first_child_index: u8,
+    num_children: u8,
+    left: Rc<BinaryTreeNode>,
+    right: Rc<BinaryTreeNode>,
+    hash: HashValue,
+}
+
+impl BinaryTreeInternalNode {
+    fn in_left_subtree(&self, n: u8) -> bool {
+        assert!(n >= self.first_child_index);
+        assert!(n < self.first_child_index + self.num_children);
+
+        n < self.first_child_index + self.num_children / 2
+    }
+}
+
+#[derive(Clone, Copy)]
+struct BinaryTreeChildNode {
+    version: Version,
+    index: u8,
```

Comment:
Would be good to add some comments for these fields. It takes some time to figure out what `first_child_index`, `num_children`, `index` are.

Reply:
confused about `index` too